apps/aehttp/test/aehttp_hyperchains_SUITE.erl

```diff
@@ -1080,11 +1080,201 @@ wallet_post_pin_to_pc(Config) ->
+           first  := _First,
+           last   := Last,
+           length := _Length}} = rpc(Node, aec_chain_hc, epoch_info, []),
+    CH = rpc(Node, aec_chain, top_height, []),
+    DistToBeforeLast = Last - CH - 1,
+    {ok, _} = produce_cc_blocks(Config, DistToBeforeLast).
+
+% PINREFAC
+pin_contract_call_tx(Config, Fun, Args, Amount, FromPubKey) ->
+    ContractPubkey = ?config(election_contract, Config),
+    Nonce = next_nonce(?NODE1, FromPubKey),
+    {ok, CallData} = aeb_fate_abi:create_calldata(Fun, Args),
+    ABI = aect_test_utils:abi_version(),
+    TxSpec =
+        #{  caller_id   => aeser_id:create(account, FromPubKey)
+          , nonce       => Nonce
+          , contract_id => aeser_id:create(contract, ContractPubkey)
+          , abi_version => ABI
+          , fee         => 1000000 * ?DEFAULT_GAS_PRICE
+          , amount      => Amount
+          , gas         => 1000000
+          , gas_price   => ?DEFAULT_GAS_PRICE
+          , call_data   => CallData},
+    {ok, Tx} = aect_call_tx:new(TxSpec),
+    NetworkId = ?config(network_id, Config),
+    SignedTx = sign_tx(Tx, privkey(who_by_pubkey(FromPubKey)), NetworkId),
+    rpc:call(?NODE1_NAME, aec_tx_pool, push, [SignedTx, tx_received]),
+    ok.
+
+% PINREFAC aec_parent_connector??
+pin_to_parent(Node, PinningData, AccountPK) ->
+    AccPKEncEnc = aeser_api_encoder:encode(account_pubkey, AccountPK),
+    {ok, []} = rpc(?PARENT_CHAIN_NODE, aec_tx_pool, peek, [infinity]), % no pending transactions
+    PinTx = rpc(Node, aec_parent_connector, create_pin_tx, [AccPKEncEnc, AccountPK, 1, 30000 * ?DEFAULT_GAS_PRICE, PinningData]),
+    SignedPinTx = sign_tx(PinTx, privkey(?DWIGHT),?PARENT_CHAIN_NETWORK_ID),
+    rpc(Node, aec_parent_connector, post_pin_tx, [SignedPinTx]).
+
+% PINREFAC
+tx_hash_to_child(Node, EncTxHash, SendAccount, Leader, Config) ->
+    NodeName = aecore_suite_utils:node_name(Node),
+    NetworkId = ?config(network_id, Config), % TODO not 100% sure about this one...
+    Nonce = next_nonce(Node, pubkey(SendAccount)),
+    Params = #{ sender_id    => aeser_id:create(account, pubkey(SendAccount)),
+                recipient_id => aeser_id:create(account, Leader),
+                amount       => 1,
+                fee          => 30000 * ?DEFAULT_GAS_PRICE,
+                nonce        => Nonce,
+                payload      => EncTxHash},
+    ct:log("Preparing a spend tx: ~p", [Params]),
+    {ok, Tx} = aec_spend_tx:new(Params),
+    SignedTx = sign_tx(Tx, privkey(SendAccount), NetworkId),
+    ok = rpc:call(NodeName, aec_tx_pool, push, [SignedTx, tx_received]),
+    Hash = rpc:call(NodeName, aetx_sign, hash, [SignedTx]),
+    Hash.
+
+mine_to_next_epoch(Node, Config) ->
+    Height1 = rpc(Node, aec_chain, top_height, []),
+    {ok, #{last := Last1, length := _Len}} = rpc(Node, aec_chain_hc, epoch_info, []),
+    {ok, Bs} = produce_cc_blocks(Config, Last1 - Height1 + 1), 
+    ct:log("Block last epoch: ~p", [hd(lists:nthtail(length(Bs) - 2, Bs))]).
```

Comment:
This does not work if you already are at last block of epoch. It will crash
Just do `ct:log` of all `Bs` to avoid that.

Reply:
Not sure I get the ct:log part of this?

Reply:
You mean we dig too deep in a list which imght be only 1 long? Is `ct:log("Block last epoch: ~p", [BS]).`your suggestion?

Reply:
I mean that if Height1 == Last1 then you only get 1 bkock and then your nth tail crashes... Which you don't want

Reply:
So, yes, your line above is a better solution

Reply:
Will go with that.